modules/xphoto/src/grayworld_white_balance.cpp

```diff
@@ -0,0 +1,232 @@
+// warranties of merchantability and fitness for a particular purpose are disclaimed.
+// In no event shall the Intel Corporation or contributors be liable for any direct,
+// indirect, incidental, special, exemplary, or consequential damages
+// (including, but not limited to, procurement of substitute goods or services;
+// loss of use, data, or profits; or business interruption) however caused
+// and on any theory of liability, whether in contract, strict liability,
+// or tort (including negligence or otherwise) arising in any way out of
+// the use of this software, even if advised of the possibility of such damage.
+//
+//M*/
+
+#include "opencv2/xphoto.hpp"
+
+#include "opencv2/core.hpp"
+#include "opencv2/hal/intrin.hpp"
+
+namespace cv { namespace xphoto {
+
+    void autowbGrayworld(InputArray _src, OutputArray _dst, const float thresh)
+    {
+
+        Mat src = _src.getMat();
+        CV_Assert(!src.empty());
+        CV_Assert(src.isContinuous());
+
+        // TODO: Handle CV_8UC1
+        // TODO: Handle types other than CV_8U
+        CV_Assert(src.type() == CV_8UC3);
+
+        _dst.create(src.size(), src.type());
+        Mat dst = _dst.getMat();
+
+        int width  = src.cols,
+            height = src.rows,
+            N      = width*height,
+            N3     = N*3;
+
+        // Calculate sum of pixel values of each channel
+        const uchar* src_data = src.ptr<uchar>(0);
+        unsigned long sum1 = 0, sum2 = 0, sum3 = 0;
+        int i = 0;
+#if CV_SIMD128
+        v_uint8x16 v_inB, v_inG, v_inR;
+        v_uint16x8 v_s1, v_s2;
+        v_uint32x4 v_iB1, v_iB2, v_iB3, v_iB4,
+                   v_iG1, v_iG2, v_iG3, v_iG4,
+                   v_iR1, v_iR2, v_iR3, v_iR4,
+                   v_SB = v_setzero_u32(),
+                   v_SG = v_setzero_u32(),
+                   v_SR = v_setzero_u32(),
+                   v_m1, v_m2, v_m3, v_m4;
+        v_float32x4 v_thresh = v_setall_f32(thresh),
+                    v_min1, v_min2, v_min3, v_min4,
+                    v_max1, v_max2, v_max3, v_max4,
+                    v_sat1, v_sat2, v_sat3, v_sat4;
+
+        for ( ; i < N3 - 47; i += 48 )
+        {
+            // NOTE: This block assumes BGR channels in naming variables
+
+            // Load 3x uint8x16 and deinterleave into vectors of each channel
+            v_load_deinterleave(&src_data[i], v_inB, v_inG, v_inR);
+
+            // Split into four int vectors per channel
+            v_expand(v_inB, v_s1, v_s2);
+            v_expand(v_s1, v_iB1, v_iB2);
+            v_expand(v_s2, v_iB3, v_iB4);
+
+            v_expand(v_inG, v_s1, v_s2);
+            v_expand(v_s1, v_iG1, v_iG2);
+            v_expand(v_s2, v_iG3, v_iG4);
+
+            v_expand(v_inR, v_s1, v_s2);
+            v_expand(v_s1, v_iR1, v_iR2);
+            v_expand(v_s2, v_iR3, v_iR4);
+
+            // Get saturation
+            v_min1 = v_cvt_f32(v_reinterpret_as_s32(v_min(v_iB1, v_min(v_iG1, v_iR1))));
```

Comment:
please, multiple thresh by 255, round it to integer and then do everything in integer arithmetics without any division

(max_ch - min_ch)/max_ch &gt;  thresh ~
round((max_ch - min_ch)*255/max_ch) &gt; round(thresh*255) ~
(max_ch - min_ch)*255 &gt; round(thresh*255)*max_ch